File: Python/other/Possible_code_recursion.py
# Given a string s, and assuming that a=1, b=2, c=3, d=4, .... z=26
# Python recursive function to print all the possible codes for the strings.


def output(original, previous):

    if len(original) == 0:  # if length is zero print 0th value i.e., '`'
        print(previous)
        return
    if len(original) == 1:
        print (previous + chr(96 + int(original)))  # checking range
        return

    if int(original[0:2]) < 27:
        new_character = chr(96 + int(original[0:2]))  # getting  new character
        output(original[2:], previous + new_character)
    new_character = chr(96 + int(original[0:1]))
    output(original[1:], previous + new_character)

# driver code
if __name__ == '__main__':
    string = input()
    output(string, '')

'''
Sample I/O:
    Input:
    1123
    Output:
    kw
    kbc
    alc
    aaw
    aabc

Time Complexity: O(N) ,where N =String
Space Complexity: O(1)
'''

File: C/cp/possible_code_recursion.c
#include <stdio.h>
#include <string.h>
#define MAX 10

void decode(char *code, char *ans, int i, int j) {
	// i is index of code array
	// j is index of ans array
	if (i >= strlen(code)) {
		printf("%s ", ans);
		return;
	}
	// getting all possible sequences recursives
	if (code[i + 1] != '0') {
		ans[j] = (code[i] - '0' + 96);
		decode(code, ans, i + 1, j + 1);
	}
	// converting a num to char
	if (code[i + 1]) {
		int result = (code[i] - '0') * 10 + (code[i + 1] - '0') - 1;
		if (result <= 25) {
			ans[j] = result + 97;
			ans[j + 1] = '\0';
			decode(code, ans, i + 2, j + 1);
		}
	}
}
int main() {
	// creating a null string
	char ans[50] = { '\0' };
	// input string
	char code[MAX];
	printf("Enter the code: ");
	scanf("%s", code);
	decode(code, ans, 0, 0);
}

/*
OUTPUT:
Enter the code: 1123
aabc aaw alc kbc kw
Time Complexity : O(n)
Space complexity : O(n)
*/
